homeassistant/components/airly/__init__.py

```diff
@@ -49,43 +53,75 @@
 
     websession = async_get_clientsession(hass)
 
-    airly = AirlyData(websession, api_key, latitude, longitude)
+    # We check how many Airly config entries are and calculate interval to not
+    # exceed allowed numbers of requests.
+    update_interval = timedelta(
+        minutes=ceil(24 * 60 / MAX_REQUESTS_PER_DAY) * instances(hass)
+    )
+
+    coordinator = AirlyDataUpdateCoordinator(
+        hass, websession, api_key, latitude, longitude, update_interval
+    )
+    await coordinator.async_refresh()
 
-    await airly.async_update()
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][config_entry.entry_id] = coordinator
 
-    hass.data[DOMAIN][DATA_CLIENT][config_entry.entry_id] = airly
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    # Change update_interval for another Airly instances
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.entry_id != config_entry.entry_id and hass.data.get(DOMAIN).get(
+            entry.entry_id
+        ):
+            hass.data.get(DOMAIN).get(entry.entry_id).update_interval = update_interval
 
-    hass.async_create_task(
-        hass.config_entries.async_forward_entry_setup(config_entry, "air_quality")
-    )
-    hass.async_create_task(
-        hass.config_entries.async_forward_entry_setup(config_entry, "sensor")
-    )
     return True
 
 
 async def async_unload_entry(hass, config_entry):
     """Unload a config entry."""
-    hass.data[DOMAIN][DATA_CLIENT].pop(config_entry.entry_id)
-    await hass.config_entries.async_forward_entry_unload(config_entry, "air_quality")
-    await hass.config_entries.async_forward_entry_unload(config_entry, "sensor")
-    return True
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(config_entry.entry_id)
+
+    # Change update_interval for another Airly instances
+    update_interval = timedelta(
+        minutes=ceil(24 * 60 / MAX_REQUESTS_PER_DAY) * (instances(hass) - 1)
+    )
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.entry_id != config_entry.entry_id and hass.data.get(DOMAIN).get(
+            entry.entry_id
+        ):
+            hass.data.get(DOMAIN).get(entry.entry_id).update_interval = update_interval
```

Comment:
Why not extract this logic so you're not duplicating it. And then have it return the update_inteval at the end for use in async_setup_entry.

Reply:
Done.